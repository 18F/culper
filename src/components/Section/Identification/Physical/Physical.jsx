import React from 'react'
import { i18n } from '../../../../config'
import schema from '../../../../schema'
import validate from '../../../../validators'
import SubsectionElement from '../../SubsectionElement'
import { Field, Height, Weight, HairColor, EyeColor, Sex } from '../../../Form'

export default class Physical extends SubsectionElement {
  constructor (props) {
    super(props)

    this.state = {
      Height: props.Height,
      Weight: props.Weight,
      HairColor: props.HairColor,
      EyeColor: props.EyeColor,
      Sex: props.Sex,
      Comments: props.Comments
    }
  }

  handleUpdate (field, values) {
    this.setState({ [field]: values }, () => {
      if (this.props.onUpdate) {
        this.props.onUpdate({
          Height: this.state.Height,
          Weight: this.state.Weight,
          HairColor: this.state.HairColor,
          EyeColor: this.state.EyeColor,
          Sex: this.state.Sex,
          Comments: this.state.Comments
        })
      }
    })
  }

  render () {
    const klass = `physical ${this.props.className || ''}`.trim()

    return (
      <div className={klass}>
        <Field title={i18n.t('identification.traits.title')}
               titleSize="h2"
               optional={true}
               className="no-margin-bottom"
               />

        <Field title={i18n.t('identification.traits.heading.height')}
               help="identification.traits.help.height"
               adjustFor='labels'
               scrollIntoView={this.props.scrollIntoView}
               shrink={true}>
          <Height name="height"
                  {...this.props.Height}
                  onUpdate={this.handleUpdate.bind(this, 'Height')}
                  onError={this.handleError}
                  required={this.props.required}
                  />
        </Field>

        <Field title={i18n.t('identification.traits.heading.weight')}
               help="identification.traits.help.weight"
               adjustFor="labels"
               scrollIntoView={this.props.scrollIntoView}
               shrink={true}>
          <Weight name="weight"
                  {...this.props.Weight}
                  onUpdate={this.handleUpdate.bind(this, 'Weight')}
                  onError={this.handleError}
                  required={this.props.required}
                  />
        </Field>

        <Field title={i18n.t('identification.traits.heading.hair')}
               adjustFor="big-buttons"
               scrollIntoView={this.props.scrollIntoView}
               help="identification.traits.help.hair">
          <HairColor name="hair"
                     help="identification.traits.help.hair"
                     className=""
                     {...this.props.HairColor}
                     onUpdate={this.handleUpdate.bind(this, 'HairColor')}
                     onError={this.handleError}
                     required={this.props.required}
                     />
        </Field>

        <Field title={i18n.t('identification.traits.heading.eye')}
               adjustFor="big-buttons"
               scrollIntoView={this.props.scrollIntoView}
               help="identification.traits.help.eye">
          <EyeColor name="eye"
                    className=""
                    {...this.props.EyeColor}
                    onUpdate={this.handleUpdate.bind(this, 'EyeColor')}
                    onError={this.handleError}
                    required={this.props.required}
                    />
        </Field>

        <Field title={i18n.t('identification.traits.heading.sex')}
               help="identification.traits.help.sex"
               adjustFor="big-buttons"
               shrink={true}
               onUpdate={this.handleUpdate.bind(this, 'Comments')}
               commentsValue={this.state.Comments}
               scrollIntoView={this.props.scrollIntoView}
               comments={true}>
          <Sex name="sex"
               {...this.props.Sex}
               onUpdate={this.handleUpdate.bind(this, 'Sex')}
               onError={this.handleError}
               required={this.props.required}
               />
        </Field>
      </div>
    )
  }
}

Physical.defaultProps = {
  Height: {},
  Weight: {},
  HairColor: '',
  EyeColor: '',
  Sex: '',
  Comments: {},
  onError: (value, arr) => { return arr },
  section: 'identification',
  subsection: 'physical',
  dispatch: () => {},
  validator: (data) => {
    return validate(schema('identification.physical', data))
  },
  required: false
}
