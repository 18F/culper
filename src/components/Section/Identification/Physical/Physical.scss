.physical {
  &.hair-colors,
  &.eye-colors,
  &.sex {
    /* increase line height for help toggle alignment */
    > .help {
      line-height: 7.8rem;

      .hair-colors,
      .eye-colors,
      .sex {
        line-height: 4.6rem;
      }
    }
  }
}
