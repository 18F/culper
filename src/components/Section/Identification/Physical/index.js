import Physical from './Physical'
export default Physical
