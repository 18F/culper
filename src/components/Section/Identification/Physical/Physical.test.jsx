import React from 'react'
import MockAdapter from 'axios-mock-adapter'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import Physical from './Physical'
import { mount } from 'enzyme'

describe('The physical attributes section', () => {
  // Setup
  const middlewares = [ thunk ]
  const mockStore = configureMockStore(middlewares)

  it('no error on first composition', () => {
    const component = mount(<Physical name="physical" />)
    expect(component.find('.pounds input').length).toEqual(1)
    expect(component.find('.feet input').length).toEqual(1)
    expect(component.find('.inches input').length).toEqual(1)
    expect(component.find('.hair-colors').length).toBeGreaterThan(0)
    expect(component.find('.eye-colors input').length).toBeGreaterThan(0)
    expect(component.find('.sex input').length).toBeGreaterThan(0)
    expect(component.find('.usa-input-error-label').length).toEqual(0)
  })

  it('loads data into physical section', () => {
    let updates = 0
    let data = {
      Comments: 'Hello',
      EyeColor: 'Black',
      HairColor: [
        'Bald'
      ],
      Height: {
        feet: 5,
        inches: 10
      },
      Sex: 'Male',
      Weight: 180,
      onUpdate: () => { updates++ }
    }
    const component = mount(<Physical name="physical" {...data} />)
    component.find('.pounds input').simulate('change')
    component.find('.feet input').simulate('change')
    component.find('.inches input').simulate('change')
    component.find('.bald input').simulate('change')
    component.find('.eye-colors input').first().simulate('change')
    component.find('.sex input').first().simulate('change')
    expect(updates).toBeGreaterThan(0)
  })
})
