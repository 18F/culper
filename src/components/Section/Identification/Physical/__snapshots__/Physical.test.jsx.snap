// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Physical component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="section-content physical"
    data-section="identification"
    data-subsection="physical"
  >
    <h1
      className="section-header"
    >
      Your identifying information
    </h1>
    <Field
      adjustFor="labels"
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.traits.help.height"
      helpActive={false}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={true}
      title="Height"
      titleSize="h4"
      validate={true}
    >
      <Height
        error={false}
        feet=""
        inches=""
        name="height"
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        tab={[Function]}
        valid={false}
      />
    </Field>
    <Field
      adjustFor="labels"
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.traits.help.weight"
      helpActive={false}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={true}
      title="Weight"
      titleSize="h4"
      validate={true}
    >
      <Weight
        name="weight"
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        value=""
      />
    </Field>
    <Field
      adjustFor="big-buttons"
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.traits.help.hair"
      helpActive={false}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={false}
      title="Hair color"
      titleSize="h4"
      validate={true}
    >
      <HairColor
        className=""
        help="identification.traits.help.hair"
        name="hair"
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        value=""
      />
    </Field>
    <Field
      adjustFor="big-buttons"
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.traits.help.eye"
      helpActive={false}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={false}
      title="Eye color"
      titleSize="h4"
      validate={true}
    >
      <EyeColor
        className=""
        name="eye"
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
      />
    </Field>
    <Field
      adjustFor="big-buttons"
      className=""
      comments={true}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.traits.help.sex"
      helpActive={false}
      onUpdate={[Function]}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={true}
      title="Select your sex"
      titleSize="h4"
      validate={true}
    >
      <Sex
        name="sex"
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        value=""
      />
    </Field>
  </div>,
  "nodes": Array [
    <div
      className="section-content physical"
      data-section="identification"
      data-subsection="physical"
    >
      <h1
        className="section-header"
      >
        Your identifying information
      </h1>
      <Field
        adjustFor="labels"
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.traits.help.height"
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={true}
        title="Height"
        titleSize="h4"
        validate={true}
      >
        <Height
          error={false}
          feet=""
          inches=""
          name="height"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          tab={[Function]}
          valid={false}
        />
      </Field>
      <Field
        adjustFor="labels"
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.traits.help.weight"
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={true}
        title="Weight"
        titleSize="h4"
        validate={true}
      >
        <Weight
          name="weight"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          value=""
        />
      </Field>
      <Field
        adjustFor="big-buttons"
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.traits.help.hair"
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Hair color"
        titleSize="h4"
        validate={true}
      >
        <HairColor
          className=""
          help="identification.traits.help.hair"
          name="hair"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          value=""
        />
      </Field>
      <Field
        adjustFor="big-buttons"
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.traits.help.eye"
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Eye color"
        titleSize="h4"
        validate={true}
      >
        <EyeColor
          className=""
          name="eye"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
        />
      </Field>
      <Field
        adjustFor="big-buttons"
        className=""
        comments={true}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.traits.help.sex"
        helpActive={false}
        onUpdate={[Function]}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={true}
        title="Select your sex"
        titleSize="h4"
        validate={true}
      >
        <Sex
          name="sex"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          value=""
        />
      </Field>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Physical
        Comments={Object {}}
        EyeColor=""
        HairColor=""
        Height={Object {}}
        Sex=""
        Weight={Object {}}
        dispatch={[Function]}
        onError={[Function]}
        required={false}
        validator={[Function]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Physical {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleCompletion": [Function],
        "handleError": [Function],
        "props": Object {
          "Comments": Object {},
          "EyeColor": "",
          "HairColor": "",
          "Height": Object {},
          "Sex": "",
          "Weight": Object {},
          "dispatch": [Function],
          "onError": [Function],
          "required": false,
          "validator": [Function],
        },
        "refs": Object {},
        "section": "identification",
        "state": null,
        "store": "Identification",
        "storeKey": "Physical",
        "subsection": "physical",
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="section-content physical"
          data-section="identification"
          data-subsection="physical"
        >
          <h1
            className="section-header"
          >
            Your identifying information
          </h1>
          <Field
            adjustFor="labels"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.height"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={true}
            title="Height"
            titleSize="h4"
            validate={true}
          >
            <Height
              error={false}
              feet=""
              inches=""
              name="height"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              tab={[Function]}
              valid={false}
            />
          </Field>
          <Field
            adjustFor="labels"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.weight"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={true}
            title="Weight"
            titleSize="h4"
            validate={true}
          >
            <Weight
              name="weight"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              value=""
            />
          </Field>
          <Field
            adjustFor="big-buttons"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.hair"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Hair color"
            titleSize="h4"
            validate={true}
          >
            <HairColor
              className=""
              help="identification.traits.help.hair"
              name="hair"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              value=""
            />
          </Field>
          <Field
            adjustFor="big-buttons"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.eye"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Eye color"
            titleSize="h4"
            validate={true}
          >
            <EyeColor
              className=""
              name="eye"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
            />
          </Field>
          <Field
            adjustFor="big-buttons"
            className=""
            comments={true}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.sex"
            helpActive={false}
            onUpdate={[Function]}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={true}
            title="Select your sex"
            titleSize="h4"
            validate={true}
          >
            <Sex
              name="sex"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              value=""
            />
          </Field>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="section-content physical"
          data-section="identification"
          data-subsection="physical"
        >
          <h1
            className="section-header"
          >
            Your identifying information
          </h1>
          <Field
            adjustFor="labels"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.height"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={true}
            title="Height"
            titleSize="h4"
            validate={true}
          >
            <Height
              error={false}
              feet=""
              inches=""
              name="height"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              tab={[Function]}
              valid={false}
            />
          </Field>
          <Field
            adjustFor="labels"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.weight"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={true}
            title="Weight"
            titleSize="h4"
            validate={true}
          >
            <Weight
              name="weight"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              value=""
            />
          </Field>
          <Field
            adjustFor="big-buttons"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.hair"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Hair color"
            titleSize="h4"
            validate={true}
          >
            <HairColor
              className=""
              help="identification.traits.help.hair"
              name="hair"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              value=""
            />
          </Field>
          <Field
            adjustFor="big-buttons"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.eye"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Eye color"
            titleSize="h4"
            validate={true}
          >
            <EyeColor
              className=""
              name="eye"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
            />
          </Field>
          <Field
            adjustFor="big-buttons"
            className=""
            comments={true}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.traits.help.sex"
            helpActive={false}
            onUpdate={[Function]}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={true}
            title="Select your sex"
            titleSize="h4"
            validate={true}
          >
            <Sex
              name="sex"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              value=""
            />
          </Field>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Physical
    Comments={Object {}}
    EyeColor=""
    HairColor=""
    Height={Object {}}
    Sex=""
    Weight={Object {}}
    dispatch={[Function]}
    onError={[Function]}
    required={false}
    validator={[Function]}
  />,
}
`;
